### PR TITLE
BIP39 Integration

### PR DESCRIPTION
**Pull request by Maneesha-rubix**
The Bitcoin Improvement Proposal number 39 (BIP39) is an implementation design for cryptocurrency wallets to generate mnemonic phrases and convert them into binary seeds, which can be used to create deterministic wallets. We are integrating BIP39 signature scheme with Rubix along with backward compatibility to NLSS signature. 

A bug with port unavailability, which was leading the nodes to fail randomly during transactions, has also been resolved in this branch.

This branch has been tested with the following test cases:
5 NLSS quorums created with BIP exe
5 BIP quorums created with BIP exe
node1, node2: NLSS node with BIP exe 
node3, node4: BIP node with BIP exe
node5: node created with old exe(main branch) with old tokens

TEST CASES:
case-1: NLSS quorums for all nodes
1. node1 >(1 rbt)> node2 >(3 rbt)> node1 >(3.5 rbt)> node3 >(0.7 rbt)> node4 >(1 rbt)> snder2 >(1.8 rbt)> node4
2. node5 >(2.6 rbt)> node4; node5 >(1 rbt)> node4; node5 >(4 rbt)> node4 >(5 rbt)> node1>(1 rbt)> node5

case-2: BIP quorums for all nodes
1. node1 >(0.2 rbt)> node2 >(1 rbt)> node1 >(2.2 rbt)> node3 >(1 rbt)> node4 >(2 rbt)> snder2 >(1 rbt)> node4
2. node5 >(1 rbt)> node1; node5 >(0.5 rbt)> node1 >(1 rbt)> node5 >(2 rbt)> node1; node5 >(1 rbt)> node3

case-3: 2 BIp quorum, 2 NLSS quorum(BIPexe), 1 NLSS quorum( old exe & old tokens : node5)
node1 >(1 rbt)> node2 >(2.1 rbt)> node1 >(1.2 rbt)> node3 >(1 rbt)> node4 >(2 rbt)> snder2 >(1 rbt)> node4

Following are the branches merged to `develop` before creating this branch :
```
  remotes/origin/ashita/explorer-url
  remotes/origin/core
  remotes/origin/core-PvtShareSigning
  remotes/origin/develop
  remotes/origin/feature/type1-quorum-selection
  remotes/origin/hari/crypto-utils
  remotes/origin/hari/determintistic-ecdsa
  remotes/origin/hari/slipkey
  remotes/origin/kiran/feature/core/PartToken
  remotes/origin/migrate
  remotes/origin/newcore
  remotes/origin/revert
```

### CODE REVIEW

**Comment by gklps** [comment]
Hi @Maneesha-rubix 
Kindly check the conflict and clear as needed please

**Comment by gklps** [comment]
Please clear the conflict @Maneesha-rubix

**Comment by Maneesha-rubix** [comment]
> Please clear the conflict @Maneesha-rubix

Conflicts resolved

File: core/ipfsport/peer.go [hunk 1]
```diff
@@ -65,13 +66,37 @@ func (pm *PeerManager) getPeerPort() uint16 {
...
+			port := pm.startPort + uint16(i)
+			availability := isPortAvailable(port)
+
+			if availability {
+				fmt.Println("available port: ", port, availability)
+				pm.ps[i] = true
+				return pm.startPort + uint16(i)
+			} else {
+				fmt.Println("NOT available port: ", port, availability)
+			}
```
Comment (gklps) [inline core/ipfsport/peer.go:78]: Is these print necessary?

Comment (Maneesha-rubix) [reply core/ipfsport/peer.go:78]: We can remove the print statements once the port unavailability is tested on MacOs.

**Comment by Maneesha-rubix** [comment]
> # Test Result:
> `The below test has been done using the same exe across all nodes on Mac. Register DID was called in sender and receiver, and testing are done in the same order as listed below`
> 
> ## Node details:
> ### NLSS
> `sender 12D3KooWKnodUB11mHeHLnHDW2vaxXqoKNwrAdFAABR4eHEhzQ7M.bafybmihnfsr2tec427ks6elys6v46wpqbhh65xlrzszqkytri7hlhebgme-20010` `receiver 12D3KooWHnPomRdXMFXHibBAZbdB9CkwtSw2VkiKabkwLSSdej12.bafybmidjnukd2nyc67wp4lv4js7guun3dz6d4a7fsz4tvtdbcdeq7t4xd4-20020`
> 
> ### BIP39
> `sender 12D3KooWKnodUB11mHeHLnHDW2vaxXqoKNwrAdFAABR4eHEhzQ7M.bafybmieox4cpc4uu3hc6jqijfd26c7xzcneuvhwywc4ze4cvdtcxclojum-20010` `receiver 12D3KooWHnPomRdXMFXHibBAZbdB9CkwtSw2VkiKabkwLSSdej12.bafybmifyf76my6at5o4rfn4sjv3gj6sbxdk76zejsof34jlv6uazsepwau-20020`
> 
> ## All quorums as NLSS
> ### Whole token
> * [x]  BIP39 as sender and NLSS as receiver
> * [ ]  NLSS as sender and BIP39 as receiver  | BIP39 as sender and BIP39 as receiver | NLSS as sender and NLSS as receiver
> 
> > ### Sender : 12D3KooWHnPomRdXMFXHibBAZbdB9CkwtSw2VkiKabkwLSSdej12.bafybmidjnukd2nyc67wp4lv4js7guun3dz6d4a7fsz4tvtdbcdeq7t4xd4 || Receiver : 12D3KooWKnodUB11mHeHLnHDW2vaxXqoKNwrAdFAABR4eHEhzQ7M.bafybmieox4cpc4uu3hc6jqijfd26c7xzcneuvhwywc4ze4cvdtcxclojum
> > Error in sender:
> > 2024-05-06T17:03:36.550+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > 2024-05-06T17:03:36.588+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > 2024-05-06T17:03:36.588+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > 2024-05-06T17:03:36.588+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:03:36.589+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:03:36.591+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > 2024-05-06T17:03:36.591+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:03:36.651+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:03:37.240+0530 [ERROR] Main.Core: Consensus failed, retry transaction after sometimes
> > 2024-05-06T17:03:37.240+0530 [ERROR] Main.Core: Consensus failed : err="consensus failed, retry transaction after sometimes"
> > Error in quorum:
> 
> > 2024-05-06T17:03:35.245+0530 [DEBUG] Main.Core: Request for pledge
> > 2024-05-06T17:03:35.245+0530 [ERROR] Main.Core: failed to get token: err="no records found"
> > 2024-05-06T17:03:36.252+0530 [DEBUG] Main.Core: RBT consensus started
> > 2024-05-06T17:03:36.462+0530 [INFO]  Main.Core: Pins help by current sender and receiver, pass
> > 2024-05-06T17:03:36.463+0530 [DEBUG] Main.Core: Token does not have multiple pins
> > available port:  25011 true
> > 2024-05-06T17:03:36.549+0530 [ERROR] Main.Core.wallet: Failed to get DID: err="no records found"
> > 2024-05-06T17:03:36.549+0530 [ERROR] Main.Core: failed to setup foreign DID quorum: err="couldn't fetch did type"
> > 2024-05-06T17:03:36.550+0530 [ERROR] Main.Core: Tokens ownership check failed
> 
> ### Parts token
> * [ ]  BIP39 as sender and NLSS as receiver | NLSS as sender and BIP39 as receiver | BIP39 as sender and BIP39 as receiver
> 
> > ### Sender : 12D3KooWKnodUB11mHeHLnHDW2vaxXqoKNwrAdFAABR4eHEhzQ7M.bafybmieox4cpc4uu3hc6jqijfd26c7xzcneuvhwywc4ze4cvdtcxclojum || Receiver : 12D3KooWHnPomRdXMFXHibBAZbdB9CkwtSw2VkiKabkwLSSdej12.bafybmidjnukd2nyc67wp4lv4js7guun3dz6d4a7fsz4tvtdbcdeq7t4xd4
> > Error in Sender:
> > 2024-05-06T16:49:08.677+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > 2024-05-06T16:49:08.683+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > 2024-05-06T16:49:08.717+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > 2024-05-06T16:49:08.717+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > 2024-05-06T16:49:08.721+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T16:49:08.721+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T16:49:08.721+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T16:49:08.837+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T16:49:09.276+0530 [ERROR] Main.Core: Consensus failed, retry transaction after sometimes
> > 2024-05-06T16:49:09.278+0530 [ERROR] Main.Core: Consensus failed : err="consensus failed, retry transaction after sometimes"
> > Error in quorum
> > 2024-05-06T16:49:08.677+0530 [ERROR] Main.Core.wallet: Failed to get DID: err="no records found"
> > 2024-05-06T16:49:08.678+0530 [ERROR] Main.Core: failed to setup foreign DID quorum: err="couldn't fetch did type"
> > 2024-05-06T16:49:08.678+0530 [ERROR] Main.Core: Tokens ownership check failed
> 
> * [ ]  NLSS as sender and NLSS as receiver
> 
> > Error in sender:
> > 2024-05-06T17:19:56.875+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:19:56.875+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:19:56.875+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:19:56.876+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:19:56.876+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:19:56.885+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:19:56.885+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:19:56.885+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:19:57.696+0530 [ERROR] Main.Core: Consensus failed, retry transaction after sometimes
> > 2024-05-06T17:19:57.697+0530 [ERROR] Main.Core: Consensus failed : err="consensus failed, retry transaction after sometimes"
> > Error in quorum:
> > 2024-05-06T17:19:55.686+0530 [INFO]  Main.Core: Checking Quorum Status
> > 2024-05-06T17:19:56.874+0530 [DEBUG] Main.Core: RBT consensus started
> > 2024-05-06T17:19:56.874+0530 [ERROR] Main.Core.wallet: Failed to get DID: err="no records found"
> > 2024-05-06T17:19:56.874+0530 [ERROR] Main.Core: Failed to get DID: err="couldn't fetch did type"
> 
> ### Mixed token
> * [ ]  BIP39 as sender and NLSS as receiver | NLSS as sender and BIP39 as receiver | BIP39 as sender and BIP39 as receiver | NLSS as sender and NLSS as receiver
> 
> > Error in sender:
> > 2024-05-06T17:28:36.067+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:28:36.069+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:28:36.235+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:28:36.236+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:28:36.236+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:28:36.236+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > 2024-05-06T17:28:36.237+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:28:36.238+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:28:36.238+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:28:36.239+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > 2024-05-06T17:28:37.056+0530 [ERROR] Main.Core: Consensus failed, retry transaction after sometimes
> > 2024-05-06T17:28:37.057+0530 [ERROR] Main.Core: Consensus failed : err="consensus failed, retry transaction after sometimes"
> > Error in quorum:
> > 2024-05-06T17:28:36.233+0530 [DEBUG] Main.Core: RBT consensus started
> > 2024-05-06T17:28:36.235+0530 [ERROR] Main.Core.wallet: Failed to get DID: err="no records found"
> > 2024-05-06T17:28:36.235+0530 [ERROR] Main.Core: Failed to get DID: err="couldn't fetch did type"
> 
> ## 3 quorums in BIP32 and rest quorums in NLSS (To be tested)
> ### Whole token
> * [ ]  BIP39 as sender and NLSS as receiver
> * [ ]  NLSS as sender and BIP39 as receiver
> * [ ]  BIP39 as sender and BIP39 as receiver
> * [ ]  NLSS as sender and NLSS as receiver
> 
> ### Parts token
> * [ ]  BIP39 as sender and NLSS as receiver
> * [ ]  NLSS as sender and BIP39 as receiver
> * [ ]  BIP39 as sender and BIP39 as receiver
> * [ ]  NLSS as sender and NLSS as receiver
> 
> ### Mixed token
> * [ ]  BIP39 as sender and NLSS as receiver
> * [ ]  NLSS as sender and BIP39 as receiver
> * [ ]  BIP39 as sender and BIP39 as receiver
> * [ ]  NLSS as sender and NLSS as receiver

In above cases of failure, the error "couldn't fetch did type" is due to not being able to find did type of the particular did in PeerDID Table. This issue arises when the did is not registered properly. So, as you suggested,  currently I'm working on connecting to the peer and fetching the did type, whenever this issue arises.

**Comment by Maneesha-rubix** [comment]
> > # Test Result:
> > `The below test has been done using the same exe across all nodes on Mac. Register DID was called in sender and receiver, and testing are done in the same order as listed below`
> > ## Node details:
> > ### NLSS
> > `sender 12D3KooWKnodUB11mHeHLnHDW2vaxXqoKNwrAdFAABR4eHEhzQ7M.bafybmihnfsr2tec427ks6elys6v46wpqbhh65xlrzszqkytri7hlhebgme-20010` `receiver 12D3KooWHnPomRdXMFXHibBAZbdB9CkwtSw2VkiKabkwLSSdej12.bafybmidjnukd2nyc67wp4lv4js7guun3dz6d4a7fsz4tvtdbcdeq7t4xd4-20020`
> > ### BIP39
> > `sender 12D3KooWKnodUB11mHeHLnHDW2vaxXqoKNwrAdFAABR4eHEhzQ7M.bafybmieox4cpc4uu3hc6jqijfd26c7xzcneuvhwywc4ze4cvdtcxclojum-20010` `receiver 12D3KooWHnPomRdXMFXHibBAZbdB9CkwtSw2VkiKabkwLSSdej12.bafybmifyf76my6at5o4rfn4sjv3gj6sbxdk76zejsof34jlv6uazsepwau-20020`
> > ## All quorums as NLSS
> > ### Whole token
> > 
> > * [x]   BIP39 as sender and NLSS as receiver
> > * [ ]   NLSS as sender and BIP39 as receiver  | BIP39 as sender and BIP39 as receiver | NLSS as sender and NLSS as receiver
> > 
> > > ### Sender : 12D3KooWHnPomRdXMFXHibBAZbdB9CkwtSw2VkiKabkwLSSdej12.bafybmidjnukd2nyc67wp4lv4js7guun3dz6d4a7fsz4tvtdbcdeq7t4xd4 || Receiver : 12D3KooWKnodUB11mHeHLnHDW2vaxXqoKNwrAdFAABR4eHEhzQ7M.bafybmieox4cpc4uu3hc6jqijfd26c7xzcneuvhwywc4ze4cvdtcxclojum
> > > Error in sender:
> > > 2024-05-06T17:03:36.550+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > > 2024-05-06T17:03:36.588+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > > 2024-05-06T17:03:36.588+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > > 2024-05-06T17:03:36.588+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:03:36.589+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:03:36.591+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > > 2024-05-06T17:03:36.591+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:03:36.651+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:03:37.240+0530 [ERROR] Main.Core: Consensus failed, retry transaction after sometimes
> > > 2024-05-06T17:03:37.240+0530 [ERROR] Main.Core: Consensus failed : err="consensus failed, retry transaction after sometimes"
> > > Error in quorum:
> > 
> > 
> > > 2024-05-06T17:03:35.245+0530 [DEBUG] Main.Core: Request for pledge
> > > 2024-05-06T17:03:35.245+0530 [ERROR] Main.Core: failed to get token: err="no records found"
> > > 2024-05-06T17:03:36.252+0530 [DEBUG] Main.Core: RBT consensus started
> > > 2024-05-06T17:03:36.462+0530 [INFO]  Main.Core: Pins help by current sender and receiver, pass
> > > 2024-05-06T17:03:36.463+0530 [DEBUG] Main.Core: Token does not have multiple pins
> > > available port:  25011 true
> > > 2024-05-06T17:03:36.549+0530 [ERROR] Main.Core.wallet: Failed to get DID: err="no records found"
> > > 2024-05-06T17:03:36.549+0530 [ERROR] Main.Core: failed to setup foreign DID quorum: err="couldn't fetch did type"
> > > 2024-05-06T17:03:36.550+0530 [ERROR] Main.Core: Tokens ownership check failed
> > 
> > 
> > ### Parts token
> > 
> > * [ ]   BIP39 as sender and NLSS as receiver | NLSS as sender and BIP39 as receiver | BIP39 as sender and BIP39 as receiver
> > 
> > > ### Sender : 12D3KooWKnodUB11mHeHLnHDW2vaxXqoKNwrAdFAABR4eHEhzQ7M.bafybmieox4cpc4uu3hc6jqijfd26c7xzcneuvhwywc4ze4cvdtcxclojum || Receiver : 12D3KooWHnPomRdXMFXHibBAZbdB9CkwtSw2VkiKabkwLSSdej12.bafybmidjnukd2nyc67wp4lv4js7guun3dz6d4a7fsz4tvtdbcdeq7t4xd4
> > > Error in Sender:
> > > 2024-05-06T16:49:08.677+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > > 2024-05-06T16:49:08.683+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > > 2024-05-06T16:49:08.717+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > > 2024-05-06T16:49:08.717+0530 [ERROR] Main.Core: Failed to get consensus: msg="Token ownership check failed, err : failed to setup foreign DID quorum : %!(EXTRA string=bafybmidxv4he2yuv45eazasb6vi4dqbn45y6lzfgz3qr3tr35hwdx2yddy, string=err, *errors.errorString=couldn't fetch did type)"
> > > 2024-05-06T16:49:08.721+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T16:49:08.721+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T16:49:08.721+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T16:49:08.837+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T16:49:09.276+0530 [ERROR] Main.Core: Consensus failed, retry transaction after sometimes
> > > 2024-05-06T16:49:09.278+0530 [ERROR] Main.Core: Consensus failed : err="consensus failed, retry transaction after sometimes"
> > > Error in quorum
> > > 2024-05-06T16:49:08.677+0530 [ERROR] Main.Core.wallet: Failed to get DID: err="no records found"
> > > 2024-05-06T16:49:08.678+0530 [ERROR] Main.Core: failed to setup foreign DID quorum: err="couldn't fetch did type"
> > > 2024-05-06T16:49:08.678+0530 [ERROR] Main.Core: Tokens ownership check failed
> > 
> > 
> > 
> > * [ ]   NLSS as sender and NLSS as receiver
> > 
> > > Error in sender:
> > > 2024-05-06T17:19:56.875+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:19:56.875+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:19:56.875+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:19:56.876+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:19:56.876+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:19:56.885+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:19:56.885+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:19:56.885+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:19:57.696+0530 [ERROR] Main.Core: Consensus failed, retry transaction after sometimes
> > > 2024-05-06T17:19:57.697+0530 [ERROR] Main.Core: Consensus failed : err="consensus failed, retry transaction after sometimes"
> > > Error in quorum:
> > > 2024-05-06T17:19:55.686+0530 [INFO]  Main.Core: Checking Quorum Status
> > > 2024-05-06T17:19:56.874+0530 [DEBUG] Main.Core: RBT consensus started
> > > 2024-05-06T17:19:56.874+0530 [ERROR] Main.Core.wallet: Failed to get DID: err="no records found"
> > > 2024-05-06T17:19:56.874+0530 [ERROR] Main.Core: Failed to get DID: err="couldn't fetch did type"
> > 
> > 
> > ### Mixed token
> > 
> > * [ ]   BIP39 as sender and NLSS as receiver | NLSS as sender and BIP39 as receiver | BIP39 as sender and BIP39 as receiver | NLSS as sender and NLSS as receiver
> > 
> > > Error in sender:
> > > 2024-05-06T17:28:36.067+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:28:36.069+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:28:36.235+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:28:36.236+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:28:36.236+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:28:36.236+0530 [ERROR] Main.Core: Failed to get consensus: msg="Failed to verify sender signature"
> > > 2024-05-06T17:28:36.237+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:28:36.238+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:28:36.238+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:28:36.239+0530 [INFO]  Main.Core.wallet: No locked tokens to release
> > > 2024-05-06T17:28:37.056+0530 [ERROR] Main.Core: Consensus failed, retry transaction after sometimes
> > > 2024-05-06T17:28:37.057+0530 [ERROR] Main.Core: Consensus failed : err="consensus failed, retry transaction after sometimes"
> > > Error in quorum:
> > > 2024-05-06T17:28:36.233+0530 [DEBUG] Main.Core: RBT consensus started
> > > 2024-05-06T17:28:36.235+0530 [ERROR] Main.Core.wallet: Failed to get DID: err="no records found"
> > > 2024-05-06T17:28:36.235+0530 [ERROR] Main.Core: Failed to get DID: err="couldn't fetch did type"
> > 
> > 
> > ## 3 quorums in BIP32 and rest quorums in NLSS (To be tested)
> > ### Whole token
> > 
> > * [ ]   BIP39 as sender and NLSS as receiver
> > * [ ]   NLSS as sender and BIP39 as receiver
> > * [ ]   BIP39 as sender and BIP39 as receiver
> > * [ ]   NLSS as sender and NLSS as receiver
> > 
> > ### Parts token
> > 
> > * [ ]   BIP39 as sender and NLSS as receiver
> > * [ ]   NLSS as sender and BIP39 as receiver
> > * [ ]   BIP39 as sender and BIP39 as receiver
> > * [ ]   NLSS as sender and NLSS as receiver
> > 
> > ### Mixed token
> > 
> > * [ ]   BIP39 as sender and NLSS as receiver
> > * [ ]   NLSS as sender and BIP39 as receiver
> > * [ ]   BIP39 as sender and BIP39 as receiver
> > * [ ]   NLSS as sender and NLSS as receiver
> 
> In above cases of failure, the error "couldn't fetch did type" is due to not being able to find did type of the particular did in PeerDID Table. This issue arises when the did is not registered properly. So, as you suggested, currently I'm working on connecting to the peer and fetching the did type, whenever this issue arises.

Issue resolved

**Review by harirubix — APPROVED** [review]

**Comment by harirubix** [comment]
Tested BIP - BIP ,  NLSS - BIP , BIP - NLSS .

**Comment by gklps** [comment]
**Calling registerdid the trnx run as expected**

Tested scenarios:

1. BIP to NLSS

- [x] Whole
- [x] Parts
- [x] Mix

2. BIP to BIP

- [x] Whole
- [x] Parts
- [x] Mix

3. NLSS to NLSS

- [x] Whole
- [x] Parts
- [x] Mix

4. NLSS to BIP

- [x] Whole
- [x] Parts
- [x] Mix

**Pull request merged**